test/unit/math/test_ad_matvar.hpp

```diff
@@ -0,0 +1,724 @@
+                            const ad_tolerances& tols,
+                            std::index_sequence<I...> i) {
+  std::vector<int> A(
+      {(expect_near_rel_matvar(
+            message + std::string(", argument ") + std::to_string(I + 1),
+            std::get<I>(x), std::get<I>(y), tols),
+        0)...});
+}
+
+/**
+ * Overload for tuples of arguments. This recursively calls
+ * `expect_near_rel_matvar` on each input pair
+ *
+ * @tparam T1 Tuple type of first argument
+ * @tparam T2 Tuple Type of second argument
+ * @param message Debug message to print on fail
+ * @param x First argument to compare
+ * @param y Second argument to compare
+ * @param tols Tolerances for comparison
+ */
+template <typename... T1, typename... T2>
+void expect_near_rel_matvar(const std::string& message,
+                            const std::tuple<T1...>& x,
+                            const std::tuple<T2...>& y,
+                            const ad_tolerances& tols) {
+  if (!(sizeof...(T1) == sizeof...(T2))) {
+    FAIL() << "The number of arguments in each tuple must match";
+  }
+
+  expect_near_rel_matvar(message, x, y, tols,
+                         std::make_index_sequence<sizeof...(T1)>());
+}
+///@}
+
+/** @name test_matvar_gradient
+ * Given the inputs and outputs of a function called with Eigen
+ * matrices of vars (matvars) and vars with inner Eigen types (varmats),
+ * check that the values of the outputs are equal and that gradients with
+ * respect to each output are the same
+ *
+ * @tparam ResultMatVar Output type of matvar calculation
+ * @tparam ResultVarMat Output type of varmat calculation
+ * @tparam MatVarArgs Argument types for matvar calculation
+ * @tparam VarMatArgs Argument types for varmat calculation
+ * @param tols Test tolerances
+ * @param A_mv_f Output for matvar calculation
+ * @param A_vm_f Output for varmat calculation
+ * @param A_mv_tuple Input arguments to matvar calculation as tuple
+ * @param A_vm_tuple Input arguments to varmat calculation as tuple
+ */
+///@{
+/**
+ * Overload for scalar outputs
+ */
+template <typename ResultMatVar, typename ResultVarMat, typename... MatVarArgs,
+          typename... VarMatArgs,
+          require_all_var_vt<std::is_arithmetic, ResultMatVar,
+                             ResultVarMat>* = nullptr>
+inline void test_matvar_gradient(const ad_tolerances& tols,
+                                 ResultMatVar& A_mv_f, ResultVarMat& A_vm_f,
+                                 const std::tuple<MatVarArgs...>& A_mv_tuple,
+                                 const std::tuple<VarMatArgs...>& A_vm_tuple) {
+  expect_near_rel("var<Matrix> vs Matrix<var> values", value_of(A_vm_f),
+                  value_of(A_mv_f), tols.gradient_val_);
+  A_vm_f.adj() = 1;
+  A_mv_f.adj() = 1;
+  stan::math::grad();
+  expect_near_rel_matvar("var<Matrix> vs Matrix<var>", A_vm_tuple, A_mv_tuple,
+                         tols);
+  stan::math::set_zero_all_adjoints();
+}
```

Comment:
Should the type names just be `ResultVar1` and `ResultVar1`? Think it's just easier to read that way

Reply:
Hmm I like keeping the VM/MV designation. `ResultVM` and `ResultMV`?

Reply:
It really is hard to read though I agree.